### PR TITLE
MV-1011: [MV-MIOpen] Benchmark & Port LPPool

### PR DESCRIPTION
**Pull request by hieule88**
* Added LPPool 1D 2D forward and backward.

* Added driver test and gtest for LPPool.

* New API is guarded by MIOPEN_BETA_API macro.

* Compared to ROCm pytorch: [Link](https://morehio-my.sharepoint.com/:x:/r/personal/kyeonghwan_ryu_moreh_io/_layouts/15/Doc.aspx?sourcedoc=%7B01B01EC0-DCD8-4E21-82E9-0F809996A3A0%7D&file=LPPool_benchmark.xlsx&action=default&mobileredirect=true)

* Average over all cases:

* LPPool 1D

|   Type    | Forward | Backward |
|----------|----------|------------|   
| float16   |   2.35    |      4.12     |
| float32   |   2.37    |      3.51     |
| bfloat16 |   2.40    |      4.17     |

* LPPool 2D

|   Type    | Forward | Backward |
|----------|----------|------------|   
| float16   |   1.25    |      1.47     |
| float32   |   1.36    |      1.68     |
| bfloat16 |   1.35    |      1.56     |

### CODE REVIEW

**Comment by littlecutebird** [comment]
please check:
 `./bin/MIOpenDriver lppool -D 512x1280x7 -k 2 -s 1 -p 2 -F 2`
`./bin/MIOpenDriver lppool -D 3x2048x64x64 -k 4x2 -s 1x1 -p 2 -F 1 -C 0`

**Comment by hieule88** [comment]
> please check: `./bin/MIOpenDriver lppool -D 512x1280x7 -k 2 -s 1 -p 2 -F 2` `./bin/MIOpenDriver lppool -D 3x2048x64x64 -k 4x2 -s 1x1 -p 2 -F 1 -C 0`

fixed it

File: include/miopen/miopen.h [hunk 1]
```diff
@@ -7791,6 +7792,77 @@ MIOPEN_EXPORT miopenStatus_t miopenPReLUBackward(miopenHandle_t handle,
...
+ * @param outputGradDesc           Tensor descriptor for output grad tensor (input)
+ * @param output_grad              Data tensor output grad (input)
+ * @param inputGradDesc            Tensor descriptor for input grad tensor (input)
+ * @param input_grad               Data tensor input grad (output)
+ * @param KD                       Kernel size in dimension D  (input)
+ * @param KH                       Kernel size in dimension H (input)
+ * @param SD                       Stride size in dimension D (input)
+ * @param SH                       Stride size in dimension H (input)
+ * @param norm_type                Norm type (input)
+ * use the number of elements in the pooling window (input)
```
Comment (anhskrttt) [inline include/miopen/miopen.h]: This line of description seems not to point to any specific input param. Is this a note or anything?

File: test/gtest/lppool.hpp [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+    {
+        double threshold = std::numeric_limits<T>::epsilon();
+        auto error       = miopen::rms_range(ref_input_grad, input_grad);
+
+        // for(int i = -10; i < -1; ++i)
+        // {
+        //     std::cout << "ref_input_grad[" << i << "] = " << ref_input_grad[i] << std::endl;
+        //     std::cout << "input_grad[" << i << "] = " << input_grad[i] << std::endl;
+        // }
+
```
Comment (cognaiger9) [inline test/gtest/lppool.hpp]: remove comments at here

File: driver/lppool_driver.hpp [hunk 3]
```diff
@@ -0,0 +1,559 @@
...
+    int status;
+    int forw = inflags.GetValueInt("forw");
+
+    for(int i = 0; i < input_sz; i++)
+    {
+        input[i] = prng::gen_A_to_B<Tgpu>(static_cast<Tgpu>(1.0f), static_cast<Tgpu>(10.0f));
+    }
+    status = input_dev->ToGPU(q, input.data());
+
+    if(forw == 2)
```
Comment (cognaiger9) [inline driver/lppool_driver.hpp:290]: in command line arguments, only `forw = 1` is specified

**Review by cognaiger9 — COMMENTED** [review]
I got error when run `MIOpenDriver lppool`, can you recheck it

File: test/gtest/lppool.cpp [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+#include "lppool.hpp"
+#include "gtest/gtest.h"
+using float16 = half_float::half;
+
+// FORWARD TEST
+using GPU_LPPool_fwd_FP32  = LPPoolTestFwd<float>;
+using GPU_LPPool_fwd_FP16  = LPPoolTestFwd<float16>;
+using GPU_LPPool_fwd_BFP16 = LPPoolTestFwd<bfloat16>;
+
+TEST_P(GPU_LPPool_fwd_FP32, LPPoolTestFwd)
```
Comment (cognaiger9) [inline test/gtest/lppool.cpp]:
```suggestion
TEST_P(GPU_LPPool_fwd_FP32, Test)
```
based on this naming convetion https://github.com/ROCm/MIOpen/wiki/GTest-development

File: test/gtest/lppool.cpp [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+{
+    RunTest();
+    Verify();
+};
+
+TEST_P(GPU_LPPool_fwd_FP16, LPPoolTestFwd)
```
Comment (cognaiger9) [inline test/gtest/lppool.cpp]: ditto

File: test/gtest/lppool.cpp [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+{
+    RunTest();
+    Verify();
+};
+
+TEST_P(GPU_LPPool_fwd_BFP16, LPPoolTestFwd)
```
Comment (cognaiger9) [inline test/gtest/lppool.cpp]: ditto

File: test/gtest/lppool.cpp [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+INSTANTIATE_TEST_SUITE_P(Smoke, GPU_LPPool_fwd_FP32, testing::ValuesIn(LPPoolTestConfigsFwd()));
+INSTANTIATE_TEST_SUITE_P(Smoke, GPU_LPPool_fwd_FP16, testing::ValuesIn(LPPoolTestConfigsFwd()));
+INSTANTIATE_TEST_SUITE_P(Smoke, GPU_LPPool_fwd_BFP16, testing::ValuesIn(LPPoolTestConfigsFwd()));
+
+// BACKWARD TEST
+using GPU_LPPool_bwd_FP32  = LPPoolTestBwd<float>;
+using GPU_LPPool_bwd_FP16  = LPPoolTestBwd<float16>;
+using GPU_LPPool_bwd_BFP16 = LPPoolTestBwd<bfloat16>;
+
+TEST_P(GPU_LPPool_bwd_FP32, LPPoolTestBwd)
```
Comment (cognaiger9) [inline test/gtest/lppool.cpp]: ditto

File: test/gtest/lppool.cpp [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+{
+    RunTest();
+    Verify();
+};
+
+TEST_P(GPU_LPPool_bwd_FP16, LPPoolTestBwd)
```
Comment (cognaiger9) [inline test/gtest/lppool.cpp]: ditto

File: test/gtest/lppool.cpp [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+{
+    RunTest();
+    Verify();
+};
+
+TEST_P(GPU_LPPool_bwd_BFP16, LPPoolTestBwd)
```
Comment (cognaiger9) [inline test/gtest/lppool.cpp]: ditto

File: src/include/miopen/lppool/problem_description.hpp [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+
+        return true;
+    }
+
+    bool IsValidDims() const
+    {
+        if(inputDesc.GetLengths().size() > 4 || inputDesc.GetLengths().size() < 3)
+        {
+            MIOPEN_THROW(miopenStatusBadParm, "LPPool: Only 3D and 4D tensors are supported.");
+        }
```
Comment (cognaiger9) [inline src/include/miopen/lppool/problem_description.hpp:88]: In this doc, I see that LPPool 1d also supports 2d tensor https://pytorch.org/docs/stable/generated/torch.nn.LPPool1d.html.

File: src/include/miopen/lppool/problem_description.hpp [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+
+    bool IsValidDims() const
+    {
+        if(inputGradDesc.GetLengths().size() > 4 || inputGradDesc.GetLengths().size() < 3)
+        {
+            MIOPEN_THROW(miopenStatusBadParm, "LPPool: Only 3D and 4D tensors are supported.");
+        }
+
+        return true;
+    }
```
Comment (cognaiger9) [inline src/include/miopen/lppool/problem_description.hpp:158]: same as forward

File: src/solver/lppool/forward_lppool_1d.cpp [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+namespace miopen {
+
+namespace solver {
+
+namespace lppool {
+
+bool LPPoolForward1d::IsApplicable(const ExecutionContext&,
+                                   const miopen::lppool::FwdProblemDescription& problem) const
+{
+    if(problem.GetInputDesc().GetNumDims() != 3 || problem.GetOutputDesc().GetNumDims() != 3)
```
Comment (cognaiger9) [inline src/solver/lppool/forward_lppool_1d.cpp]: Only need to check either inputDesc or outputDesc cause they have equal dim

File: src/lppool/problem_description.cpp [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+    auto output_stride = outputDesc.GetStrides();
+
+    auto input_dtype = inputDesc.GetType();
+
+    std::ostringstream ss;
+
+    ss << "lppool_fwd";
+    ss << "-input_dtype" << input_dtype;
+    ss << "-Is" << input_size;
+    ss << "-Os" << output_size;
```
Comment (cognaiger9) [inline src/lppool/problem_description.cpp]: For reducing kernel cache string: only need to include number of dimension, total element size of input at here

File: src/lppool/problem_description.cpp [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+    auto output_grad_stride = outputGradDesc.GetStrides();
+
+    auto input_dtype = inputGradDesc.GetType();
+
+    std::ostringstream ss;
+
+    ss << "lppool_bwd";
+    ss << "-input_dtype" << input_dtype;
+    ss << "-dIs" << input_grad_size;
+    ss << "-dOs" << output_grad_size;
```
Comment (cognaiger9) [inline src/lppool/problem_description.cpp]: ditto

File: test/cpu_lppool.hpp [hunk 15]
```diff
@@ -0,0 +1,247 @@
...
+ *******************************************************************************/
+#pragma once
+
+#include "tensor_holder.hpp"
+#include <miopen/tensor_view_utils.hpp>
+#include "ford.hpp"
+
+template <class T>
+void cpu_lppool_forward_1d(tensor<T> input,
+                           tensor<T>& output,
```
Comment (cognaiger9) [inline test/cpu_lppool.hpp]:
```suggestion
void cpu_lppool_forward_1d(const tensor<T>& input,
                           tensor<T>& output,
```

File: test/cpu_lppool.hpp [hunk 16]
```diff
@@ -0,0 +1,247 @@
...
+                           int64_t N,
+                           int64_t C,
+                           int64_t D,
+                           int64_t OD,
+                           tensor<int64_t> ksize,
+                           tensor<int64_t> stride,
```
Comment (cognaiger9) [inline test/cpu_lppool.hpp]:
```suggestion
                           const tensor<int64_t>& ksize,
                           const tensor<int64_t>& stride,
```

File: test/cpu_lppool.hpp [hunk 17]
```diff
@@ -0,0 +1,247 @@
...
+                           tensor<T>& output,
+                           int64_t N,
+                           int64_t C,
+                           int64_t H,
+                           int64_t W,
+                           int64_t OH,
+                           int64_t OW,
+                           tensor<int64_t> ksize,
+                           tensor<int64_t> stride,
+                           float norm_type)
```
Comment (cognaiger9) [inline test/cpu_lppool.hpp]: same as above

File: test/cpu_lppool.hpp [hunk 18]
```diff
@@ -0,0 +1,247 @@
...
+                            tensor<T> output_grad,
+                            tensor<T>& input_grad,
+                            int64_t N,
+                            int64_t C,
+                            int64_t D,
+                            int64_t OD,
+                            tensor<int64_t> ksize,
+                            tensor<int64_t> stride,
+                            float norm_type)
+{
```
Comment (cognaiger9) [inline test/cpu_lppool.hpp]: same as above

File: test/cpu_lppool.hpp [hunk 19]
```diff
@@ -0,0 +1,247 @@
...
+                            tensor<T>& input_grad,
+                            int64_t N,
+                            int64_t C,
+                            int64_t H,
+                            int64_t W,
+                            int64_t OH,
+                            int64_t OW,
+                            tensor<int64_t> ksize,
+                            tensor<int64_t> stride,
+                            float norm_type)
```
Comment (cognaiger9) [inline test/cpu_lppool.hpp]: same as above

File: test/gtest/lppool.hpp [hunk 20]
```diff
@@ -0,0 +1,338 @@
...
+    LPPoolTestCase lppool_config;
+
+    tensor<T> input;
+    tensor<T> output;
+    tensor<T> ref_output;
+    tensor<int64_t> ksize;
+    tensor<int64_t> stride;
+
+    float norm_type;
+    int64_t N = 1, C = 1, D = 1, H = 1, OD = 1, OH = 1;
```
Comment (cognaiger9) [inline test/gtest/lppool.hpp:198]: should be `size_t` as in_dim is `vector<size_t>`

File: test/gtest/lppool.hpp [hunk 21]
```diff
@@ -0,0 +1,338 @@
...
+    tensor<T> input;
+    tensor<T> output;
+    tensor<T> output_grad;
+    tensor<T> input_grad;
+    tensor<T> ref_input_grad;
+    tensor<int64_t> ksize;
+    tensor<int64_t> stride;
+
+    float norm_type;
+    int64_t N, C, D, H, OD, OH;
```
Comment (cognaiger9) [inline test/gtest/lppool.hpp:326]: ditto

File: src/solver/lppool/forward_lppool_2d.cpp [hunk 22]
```diff
@@ -0,0 +1,137 @@
...
+{
+    if(!problem.IsAllContiguous())
+        return true;
+    return false;
+}
+
+bool LPPoolForward2d::IsApplicable(const ExecutionContext&,
+                                   const miopen::lppool::FwdProblemDescription& problem) const
+{
+    if(problem.GetInputDesc().GetNumDims() != 4 || problem.GetOutputDesc().GetNumDims() != 4)
```
Comment (cognaiger9) [inline src/solver/lppool/forward_lppool_2d.cpp]: ditto

File: include/miopen/miopen.h (hunk 1 above)
Comment (hieule88) [reply include/miopen/miopen.h]: removed

File: test/gtest/lppool.hpp (hunk 2 above)
Comment (hieule88) [reply test/gtest/lppool.hpp]: resolved

File: src/include/miopen/lppool/problem_description.hpp (hunk 10 above)
Comment (hieule88) [reply src/include/miopen/lppool/problem_description.hpp:88]: in torch doc, they do not have N and C counted, so 1D and 2D are actually 3D and 4D

File: src/include/miopen/lppool/problem_description.hpp (hunk 11 above)
Comment (hieule88) [reply src/include/miopen/lppool/problem_description.hpp:158]: same

File: src/solver/lppool/forward_lppool_1d.cpp (hunk 12 above)
Comment (hieule88) [reply src/solver/lppool/forward_lppool_1d.cpp]: resolved

File: test/cpu_lppool.hpp (hunk 15 above)
Comment (hieule88) [reply test/cpu_lppool.hpp]: resolved

File: test/cpu_lppool.hpp (hunk 16 above)
Comment (hieule88) [reply test/cpu_lppool.hpp]: resolved

File: test/cpu_lppool.hpp (hunk 17 above)
Comment (hieule88) [reply test/cpu_lppool.hpp]: resolved

File: test/cpu_lppool.hpp (hunk 18 above)
Comment (hieule88) [reply test/cpu_lppool.hpp]: resolved

File: test/cpu_lppool.hpp (hunk 19 above)
Comment (hieule88) [reply test/cpu_lppool.hpp]: resolved

File: src/solver/lppool/forward_lppool_2d.cpp (hunk 22 above)
Comment (hieule88) [reply src/solver/lppool/forward_lppool_2d.cpp]: resolved

File: test/gtest/lppool.cpp (hunk 4 above)
Comment (hieule88) [reply test/gtest/lppool.cpp]: resolved

File: test/gtest/lppool.cpp (hunk 5 above)
Comment (hieule88) [reply test/gtest/lppool.cpp]: resolved

File: test/gtest/lppool.cpp (hunk 6 above)
Comment (hieule88) [reply test/gtest/lppool.cpp]: resolved

File: test/gtest/lppool.cpp (hunk 7 above)
Comment (hieule88) [reply test/gtest/lppool.cpp]: resolved

File: test/gtest/lppool.cpp (hunk 8 above)
Comment (hieule88) [reply test/gtest/lppool.cpp]: resolved

File: test/gtest/lppool.cpp (hunk 9 above)
Comment (hieule88) [reply test/gtest/lppool.cpp]: resolved

File: src/lppool/problem_description.cpp (hunk 13 above)
Comment (hieule88) [reply src/lppool/problem_description.cpp]: isAllContiguous is needed for IsApplicable, removed unused

File: src/lppool/problem_description.cpp (hunk 14 above)
Comment (hieule88) [reply src/lppool/problem_description.cpp]: removed unused

File: driver/lppool_driver.hpp (hunk 3 above)
Comment (hieule88) [reply driver/lppool_driver.hpp:290]: I dont think thats a problem, many merged ops that have both forward and backward option and have only forw = 1 is specified in command line arguments

File: test/gtest/lppool.hpp (hunk 20 above)
Comment (hieule88) [reply test/gtest/lppool.hpp:198]: they should be int64_t because kernel code use them as int64_t

File: test/gtest/lppool.hpp (hunk 21 above)
Comment (hieule88) [reply test/gtest/lppool.hpp:326]: same

**Comment by hieule88** [comment]
> I got error when run `MIOpenDriver lppool`, can you recheck it

I dont get any error when run `MIOpenDriver lppool`, please give me more information of the error

**Comment by cognaiger9** [comment]
when I run `MIOpenDriver lppool`,the following error appear
```
error: non-constant-expression cannot be narrowed from type 'int64_t' (aka 'long long') to 'unsigned long long' in initializer list [-Wc++11-narrowing]
   63 |             pow(CVT_FLOAT2ACCUM(input[input_tv.get_tensor_view_idx({n, c, d})]),
```

**Comment by hieule88** [comment]
> when I run `MIOpenDriver lppool`,the following error appear
> 
> ```
> error: non-constant-expression cannot be narrowed from type 'int64_t' (aka 'long long') to 'unsigned long long' in initializer list [-Wc++11-narrowing]
>    63 |             pow(CVT_FLOAT2ACCUM(input[input_tv.get_tensor_view_idx({n, c, d})]),
> ```

please check your environment, my code run well with `MIOpenDriver lppool`
image

**Comment by cognaiger9** [comment]
LGTM
